### ISSUE TITLE
Question about multiple optimizers and lr schedulers?

### ISSUE THREAD

**Issue by xiadingZ** [williamFalcon/pytorch-lightning#263]
To train GAN,  I return two optimizers and two schudulers, 
1. in `def training_step(self, batch, batch_nb, optimizer_i):`, I return G loss if `optimizer_i == 0` else D loss if `optimizer_i == 1`. Is this correct? Does trainer use same `batch` data to call `train_step` twice, one for `optimizer_i = 0`, one for `optimizer_i=1`? 
2. If I want to call `scheduler_0.step()` if `optimizer_i == 0`, `scheduler_1.step()` if `optimizer_i == 1`, how to do?

**Comment by williamFalcon** [comment]
https://github.com/williamFalcon/pytorch-lightning/blob/master/examples/templates/gan.py

**Comment by awaelchli** [comment]
@xiadingZ 
2. is done by PL automatically. You don't have to manually call step() on the scheduler.

**Comment by xiadingZ** [comment]
I have read this templates.
But if I only want to call scheduler_0.step() when optimizer_0 is used, how to do? In docs, only have examples with two optimizers and one scheduler.

**Comment by awaelchli** [comment]
In this case I see two options:
a) Return two schedulers: regular one for optimizer 0 as you said and a dummy one that does effectively no scheduling (step is no op).
b) return no schedulers (just the optimizers) and define your scheduler outside (e.g. in your init method). Then perform the step manually when optimizer_idx == 0.

**Comment by williamFalcon** [comment]
As @awaelchli mentioned, you don’t have to. PL does it for you.

**Comment by xiadingZ** [comment]
what i want is to use two scheduler corresponding to two optimizers. In template, optimizer_0 takes effects if optimizer_i = 0, but PL call two schedulers' step(), which affects the optimizer_1.

**Comment by xiadingZ** [comment]
Thanks, I have read the source code and understood it
